src/main/java/com/amazon/opendistroforelasticsearch/search/asynchronous/service/AsynchronousSearchService.java

```diff
@@ -370,21 +366,51 @@ private void cancelAndFreeActiveAndPersistedContext(AsynchronousSearchActiveCont
+                                    //no onCancelled hook due to failure, so we invoke free active context
+                                    groupedDeletionListener.onResponse(freeActiveContext(asynchronousSearchContext));
+                                }), () -> groupedDeletionListener.onResponse(freeActiveContext(asynchronousSearchContext)));
                         logger.debug("Deleting asynchronous search id [{}] from system index ",
                                 asynchronousSearchContext.getAsynchronousSearchId());
                         persistenceService.deleteResponse(asynchronousSearchContext.getAsynchronousSearchId(), user, translatedListener);
-                    } else {
-                        cancelTask(asynchronousSearchContext, cancelTaskReason, () -> {
-                            if (response) {
+                    } else { //keep on completion is false. simply cancel task and clean up active context
+                        Runnable freeContextAction = () -> {
+                            if (freeActiveContext(asynchronousSearchContext)) {
                                 releasableListener.onResponse(true);
                             } else {
                                 releasableListener.onFailure(new ResourceNotFoundException(
                                         asynchronousSearchContext.getAsynchronousSearchId()));
                             }
-                        });
+                        };
+                        cancelTask(asynchronousSearchContext, cancelTaskReason, wrap(cancelTasksResponse -> {
+                                    logger.debug("Successfully cancelled tasks [{}] with asynchronous search [{}] with response [{}]",
+                                            asynchronousSearchContext.getTask(), asynchronousSearchContext.getAsynchronousSearchId(),
+                                            cancelTasksResponse);
+                                    /* we don't free active context here as AsynchronousSearchTask#onCancelled() takes care
+                                       of that. This ensures that freeActiveContext() is invoked only after task is cancelled or
+                                       completed */
+                                    releasableListener.onResponse(true);
+                                },
+                                e -> {
+                                    logger.error(() -> new ParameterizedMessage(
+                                            "Failed to cancel task [{}] with asynchronous search [{}] with exception",
+                                            asynchronousSearchContext.getTask(), asynchronousSearchContext.getAsynchronousSearchId()), e);
+                                    //no onCancelled hook due to failure, so we invoke free active context
+                                    freeContextAction.run();
+                                }), freeContextAction);
```

Comment:
Redundant code

Reply:
added a TODO to extract out this feature and add functional tests on free context. + cancel tasks behaviour in various scenarios